File: Universal Windows Platform Solution/Phantom of the West/Phantom of the West/Visual Novel/PotWVN.cs
using Phantom_of_the_West.Data_Management;
using Phantom_of_the_West.Visual_Novel.Chapters;
using Phantom_of_the_West.Visual_Novel.Serialization;
using Phantom_of_the_West.Visual_Novel.Serialization.Event_Flagging;
using System;
using System.Threading.Tasks;

namespace Phantom_of_the_West.Visual_Novel
{
	public class PotWVN : IVisualNovel, IObservable<IVisualNovel>
	{
		public static PotWVN MainVN
		{
			get;
			private set;
		} = new PotWVN();

		internal EventFlagsCollection EventFlags
		{
			get;
			private set;
		} = null;

		private ObservableVisualNovel observable = new ObservableVisualNovel();

		private IStoryChoiceList currentChoices = null;

		private int id = 0;

		public IStoryChoiceList CurrentChoices
		{
			get
			{
				return currentChoices;
			}
			internal set
			{
				currentChoices = value;
				NotifyObservers();
			}
		}

		public async Task StartUp()
		{
			GameState autoSave = await LoadAutoSave();
			if (autoSave != null)
			{
				LoadGame(autoSave);
			}
			else
			{
				PlayTutorial();
			}
		}

		public void NewGame()
		{
			EventFlags = new EventFlagsCollection();
			GoToState(1001);
		}

		public void PlayTutorial()
		{
			EventFlags = new EventFlagsCollection();
			GoToState(1);
		}

		public GameState SaveGame()
		{
			GameState gs = new GameState(EventFlags, id);
			return gs;
		}

		public void LoadGame(GameState gs)
		{
			EventFlags = gs.EventFlags;
			GoToState(gs.ID);
		}

		public IDisposable Subscribe(IObserver<IVisualNovel> observer)
		{
			return observable.Subscribe(observer);
		}

		internal void GoToState(int id)
		{
			this.id = id;
			CurrentChoices = ChapterSelector.GoToState(id);
		}

		private void NotifyObservers()
		{
			AutoSave();
			observable.NotifyObservers(this);
			PlayAudio();
		}

		private void AutoSave()
		{
			GameState gs = SaveGame();
			Task t = DataManager.AutoSave(gs);
		}

		private async Task<GameState> LoadAutoSave()
		{
			GameState gs = await DataManager.LoadAutoSave();
			return gs;
		}

		private void PlayAudio()
		{
			if (currentChoices != null)
			{
				currentChoices.PlayAudio();
			}
		}
	}
}

File: Android Project/app/src/main/java/com/christopherraleigh/phantomofthewest/visual_novel/PotWVN.java
package com.christopherraleigh.phantomofthewest.visual_novel;

import android.content.Context;

import com.christopherraleigh.phantomofthewest.data_management.DataManager;
import com.christopherraleigh.phantomofthewest.visual_novel.chapters.ChapterSelector;
import com.christopherraleigh.phantomofthewest.visual_novel.serialization.GameState;
import com.christopherraleigh.phantomofthewest.visual_novel.serialization.event_flagging.EventFlagsCollection;

import java.lang.ref.WeakReference;
import java.util.Observable;

/**
 * Created by christopherraleigh on 2016-04-01.
 */
public class PotWVN extends Observable {

    private static final int STORY_START = 1001;
    private static final int TUTORIAL_START = 1;
    private static PotWVN mainVN = null;
    private WeakReference<Context> context;
    private IStoryChoiceList currentChoices = null;
    private EventFlagsCollection eventFlags = null;
    private int id = 0;

    private PotWVN(Context context) {
        this.context = new WeakReference<>(context);
    }

    public static PotWVN getMainVN() {
        return mainVN;
    }

    public static void startUp(Context c) {
        mainVN = new PotWVN(c);
        GameState autoSave = mainVN.loadAutoSave();
        if (autoSave != null) {
            mainVN.loadGame(autoSave);
        } else {
            mainVN.playTutorial();
        }
    }

    public IStoryChoiceList getCurrentChoices() {
        return currentChoices;
    }

    private void setCurrentChoices(IStoryChoiceList choices) {
        currentChoices = choices;
        prepareToNotifyObservers();
    }

    public EventFlagsCollection getEventFlags() {
        return eventFlags;
    }

    public void goToState(int id) {
        this.id = id;
        setCurrentChoices(ChapterSelector.goToState(id));
    }

    public void loadGame(GameState gs) {
        if (gs != null) {
            eventFlags = gs.getEventFlags();
            goToState(gs.getID());
        }
    }

    public void newGame() {
        newGameAt(STORY_START);
    }

    public void playTutorial() {

        newGameAt(TUTORIAL_START);
    }

    public GameState saveGame() {
        GameState gs = new GameState(eventFlags, id);
        return gs;
    }

    private void autoSave() {
        Context strongContext = context.get();
        if (strongContext != null) {
            GameState gs = saveGame();
            DataManager.autoSave(strongContext, gs);
        }
    }

    private GameState loadAutoSave() {
        GameState gs = null;
        Context strongContext = context.get();
        if (strongContext != null) {
            gs = DataManager.loadAutoSave(strongContext);
        }
        return gs;
    }

    private void newGameAt(int startingPage) {
        eventFlags = new EventFlagsCollection();
        goToState(startingPage);
    }

    private void prepareToNotifyObservers() {
        autoSave();
        setChanged();
        notifyObservers(this);
    }
}